sdk/eventhub/event-hubs/src/eventHubReceiver.ts

```diff
@@ -676,153 +523,224 @@ export class EventHubReceiver extends LinkEntity {
-        // if this consumer was closed,
-        // resolve the operation's promise with the events collected thus far in case
-        // the promise hasn't already been resolved.
-        if (this._isClosed || this._context.wasConnectionCloseCalled) {
-          return resolve(receivedEvents);
-        }
-
-        let timer: any;
-        const logOnAbort = (): void => {
-          const name = this.name;
-          const address = this.address;
-          const desc: string =
-            `[${this._context.connectionId}] The request operation on the Receiver "${name}" with ` +
-            `address "${address}" has been cancelled by the user.`;
-          // Cancellation is intentional so logging to 'info'.
-          logger.info(desc);
-        };
-
-        const rejectOnAbort = async (): Promise<void> => {
-          logOnAbort();
-          try {
-            await this.close();
-          } finally {
-            reject(new AbortError(StandardAbortMessage));
-          }
-        };
+    this._isReceivingMessages = true;
+    this._isStreaming = false;
 
-        // operation has been cancelled, so exit immediately
-        if (abortSignal && abortSignal.aborted) {
-          return rejectOnAbort();
-        }
+    const cleanupBeforeAbort = (): Promise<void> => {
+      logger.info(
+        `[${this._context.connectionId}] The request operation on the Receiver "${this.name}" with address "${this.address}" has been cancelled by the user.`
+      );
+      return this.close();
+    };
 
-        // updates the prefetch count so that the baseConsumer adds
-        // the correct number of credits to receive the same number of events.
-        const prefetchCount = Math.max(maxMessageCount - receivedEvents.length, 0);
-        if (prefetchCount === 0) {
-          return resolve(receivedEvents);
+    const tryOpenLink = (): Promise<void> =>
+      createAbortablePromise(
+        (resolve, reject) =>
+          this.initialize({
+            abortSignal,
+            timeoutInMs: getRetryAttemptTimeoutInMs(this.options.retryOptions),
+          })
+            .then(resolve)
+            .catch(reject),
+        {
+          abortSignal,
+          abortErrorMsg: StandardAbortMessage,
         }
+      );
 
-        logger.verbose(
-          "[%s] Receiver '%s', setting the prefetch count to %d.",
-          this._context.connectionId,
-          this.name,
-          prefetchCount
-        );
-
-        const cleanUpBeforeReturn = (): void => {
-          this.clearHandlers();
-          clearTimeout(timer);
-        };
+    /** The time to wait in ms before attempting to read from the queue */
+    const readIntervalWaitTimeInMs = 20;
 
-        const onAbort = (): void => {
-          clearTimeout(timer);
-          rejectOnAbort();
-        };
+    const retrieveEvents = (): Promise<ReceivedEventData[]> => {
+      const eventsToRetrieveCount = Math.max(maxMessageCount - this.queue.length, 0);
```

Comment:
Curious to understand how this might affect the service-bus receiving.

Reply:
I am curious too and can take a look.

Reply:
I had a brief look. SB only uses one array. It sill wait until either max # received, or max wait time elapses. Since there's no issues reported we can keep SB as it is currently. Both EH and SB had `receiveBatch()` but could be that at some point the two packages diveraged.